<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>0.1 scRNA-seq Dataset Integration | 09-Integration.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="0.1 scRNA-seq Dataset Integration | 09-Integration.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="0.1 scRNA-seq Dataset Integration | 09-Integration.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  


<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<!-- for Facebook -->  
<meta property="og:url" content="http://hemberg-lab.github.io/scRNA.seq.course/" />
<meta property="og:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="Analysis of single-cell RNA-seq data" />
<meta name="twitter:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-71525309-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-71525309-1');
</script>

<!-- For Logo -->
<script>
  $(document).ready(function(){
    $(".book-header").prepend('<img id="SangerLogo" src="figures/sanger-logo-header.png" />')
  }
</script>

<script>
  $(document).ready(function(){
    $('.book-header').wrap( $('<a>').attr('href', 'https://www.sanger.ac.uk/').attr('target', '_blank'))
  });
</script>




<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="0.1" data-path=""><a href="#scrna-seq-dataset-integration"><i class="fa fa-check"></i><b>0.1</b> scRNA-seq Dataset Integration</a>
<ul>
<li class="chapter" data-level="0.1.1" data-path=""><a href="#introduction"><i class="fa fa-check"></i><b>0.1.1</b> Introduction</a></li>
<li class="chapter" data-level="0.1.2" data-path=""><a href="#mnn-based-methods"><i class="fa fa-check"></i><b>0.1.2</b> MNN-based methods</a></li>
<li class="chapter" data-level="0.1.3" data-path=""><a href="#cannonical-correlation-analysis-seurat-v3"><i class="fa fa-check"></i><b>0.1.3</b> Cannonical Correlation Analysis (Seurat v3)</a></li>
<li class="chapter" data-level="0.1.4" data-path=""><a href="#scrna-seq-integration-tools-benchmarks"><i class="fa fa-check"></i><b>0.1.4</b> scRNA-seq Integration Tools Benchmarks</a></li>
<li class="chapter" data-level="0.1.5" data-path=""><a href="#practical-examples-of-real-dataset-integration"><i class="fa fa-check"></i><b>0.1.5</b> Practical Examples of Real Dataset Integration</a></li>
<li class="chapter" data-level="0.1.6" data-path=""><a href="#integration-example-1-10k-pbmc-cells-profiled-using-3-and-5-10x-genomics-kits"><i class="fa fa-check"></i><b>0.1.6</b> Integration example 1: 10k PBMC cells profiled using 3’ and 5’ 10x Genomics kits</a></li>
<li class="chapter" data-level="0.1.7" data-path=""><a href="#dataset-2-10k-pbmc-cells-3-10x-and-sorted-whole-blood-generated-using-strt-seq"><i class="fa fa-check"></i><b>0.1.7</b> DATASET 2: 10k PBMC cells (3’ 10x) and sorted whole blood generated using STRT-Seq</a></li>
<li class="chapter" data-level="0.1.8" data-path=""><a href="#sessioninfo"><i class="fa fa-check"></i><b>0.1.8</b> sessionInfo()</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:title:end-->
<!--bookdown:title:start-->
<div id="scrna-seq-dataset-integration" class="section level2" number="0.1">
<h2><span class="header-section-number">0.1</span> scRNA-seq Dataset Integration</h2>
<div id="introduction" class="section level3" number="0.1.1">
<h3><span class="header-section-number">0.1.1</span> Introduction</h3>
<p>As more and more scRNA-seq datasets become available, carrying merged_seurat comparisons between them is key. There are two main approaches to comparing scRNASeq datasets. The first approach is “label-centric” which is focused on trying to identify equivalent cell-types/states across datasets by comparing individual cells or groups of cells. The other approach is “cross-dataset normalization” which attempts to computationally remove experiment-specific technical/biological effects so that data from multiple experiments can be combined and jointly analyzed.</p>
<p>The label-centric approach can be used with dataset with high-confidence cell-annotations, e.g. the Human Cell Atlas (HCA) <span class="citation">[@Regev2017-mw]</span> or the Tabula Muris <span class="citation">[@Quake2017]</span> once they are completed, to project cells or clusters from a new sample onto this reference to consider tissue composition and/or identify cells with novel/unknown identity. Conceptually, such projections are similar to the popular BLAST method <span class="citation">[@Altschul1990-ts]</span>, which makes it possible to quickly find the closest match in a database for a newly identified nucleotide or amino acid sequence. The label-centric approach can also be used to compare datasets of similar biological origin collected by different labs to ensure that the annotation and the analysis is consistent.</p>
<div class="figure" style="text-align: center"><span id="fig:int1"></span>
<img src="figures/CourseCompareTypes.png" alt="Label-centric dataset comparison can be used to compare the annotations of two different samples." width="950" />
<p class="caption">
Figure 1: Label-centric dataset comparison can be used to compare the annotations of two different samples.
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:int2"></span>
<img src="figures/CourseAtlasAssignment.png" alt="Label-centric dataset comparison can project cells from a new experiment onto an annotated reference." width="871" />
<p class="caption">
Figure 2: Label-centric dataset comparison can project cells from a new experiment onto an annotated reference.
</p>
</div>
<p>The cross-dataset normalization approach can also be used to compare datasets of similar biological origin, unlike the label-centric approach it enables the join analysis of multiple datasets to facilitate the identification of rare cell-types which may to too sparsely sampled in each individual dataset to be reliably detected. However, cross-dataset normalization is not applicable to very large and diverse references since it assumes a significant portion of the biological variablility in each of the datasets overlaps with others.</p>
<div class="figure" style="text-align: center"><span id="fig:int3"></span>
<img src="figures/CourseCrossNorm.png" alt="Cross-dataset normalization enables joint-analysis of 2+ scRNASeq datasets." width="822" />
<p class="caption">
Figure 3: Cross-dataset normalization enables joint-analysis of 2+ scRNASeq datasets.
</p>
</div>
</div>
<div id="mnn-based-methods" class="section level3" number="0.1.2">
<h3><span class="header-section-number">0.1.2</span> MNN-based methods</h3>
<p><a href="https://www.nature.com/articles/nbt.4091">mnnCorrect</a> corrects datasets to facilitate joint analysis. It order to account for differences in composition between two replicates or two different experiments it first matches invidual cells across experiments to find the overlaping biologicial structure. Using that overlap it learns which dimensions of expression correspond to the biological state and which dimensions correspond to batch/experiment effect; mnnCorrect assumes these dimensions are orthologal to each other in high dimensional expression space. Finally it removes the batch/experiment effects from the entire expression matrix to return the corrected matrix.</p>
<p>To match individual cells to each other across datasets, mnnCorrect uses the cosine distance to avoid library-size effect then identifies mututal nearest neighbours (<code>k</code> determines to neighbourhood size) across datasets. Only overlaping biological groups should have mutual nearest neighbours (see panel b below). However, this assumes that k is set to approximately the size of the smallest biological group in the datasets, but a k that is too low will identify too few mutual nearest-neighbour pairs to get a good estimate of the batch effect we want to remove.</p>
<p>Learning the biological/techncial effects is done with either singular value decomposition, similar to RUV we encounters in the batch-correction section, or with principal component analysis with the opitimized irlba package, which should be faster than SVD. The parameter <code>svd.dim</code> specifies how many dimensions should be kept to summarize the biological structure of the data, we will set it to three as we found three major groups using Metaneighbor above. These estimates may be futher adjusted by smoothing (<code>sigma</code>) and/or variance adjustment (<code>var.adj</code>).</p>
<div class="figure" style="text-align: center"><span id="fig:int4"></span>
<img src="figures/MNN.png" alt="Cross-dataset normalization enables joint-analysis of 2+ scRNASeq datasets." width="384" />
<p class="caption">
Figure 4: Cross-dataset normalization enables joint-analysis of 2+ scRNASeq datasets.
</p>
</div>
</div>
<div id="cannonical-correlation-analysis-seurat-v3" class="section level3" number="0.1.3">
<h3><span class="header-section-number">0.1.3</span> Cannonical Correlation Analysis (Seurat v3)</h3>
<p>The <code>Seurat</code> package contains another correction method for combining multiple datasets, called <a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8?">CCA</a>. However, unlike <code>mnnCorrect</code> it doesn’t correct the expression matrix itself directly. Instead <code>Seurat</code> finds a lower dimensional subspace for each dataset then corrects these subspaces. Also different from <code>mnnCorrect</code>, <code>Seurat</code> only combines a single pair of datasets at a time.</p>
<p>Seurat uses gene-gene correlations to identify the biological structure in the dataset with a method called canonical correlation analysis (CCA). Seurat learns the shared structure to the gene-gene correlations and then evaluates how well each cell fits this structure. Cells which must better described by a data-specific dimensionality reduction method than by the shared correlation structure are assumed to represent dataset-specific cell-types/states and are discarded before aligning the two datasets. Finally the two datasets are aligned using ‘warping’ algorithms which normalize the low-dimensional representations of each dataset in a way that is robust to differences in population density.</p>
</div>
<div id="scrna-seq-integration-tools-benchmarks" class="section level3" number="0.1.4">
<h3><span class="header-section-number">0.1.4</span> scRNA-seq Integration Tools Benchmarks</h3>
<p>There are several benchmarks published recently (). One of the most detailed publications compared 14 methods of scRNA-seq dataset integration using multiple simulated and real datasets of various size and complexity.</p>
<p>According to the benchmark, <code>Harmony</code>, <code>LIGER</code> (that more recently became <code>rliger</code>), and <code>Seurat</code> (v3) have performed best. We will illustrate the performance of these three methods in two tasks: 1) integrating closely related 3’ and 5’ PBMC datasets; 2) integrating only partially overlapping datasets, namely whole blood (including erythrocytes and neutrophils), and 3’ PBMC dataset.</p>
</div>
<div id="practical-examples-of-real-dataset-integration" class="section level3" number="0.1.5">
<h3><span class="header-section-number">0.1.5</span> Practical Examples of Real Dataset Integration</h3>
<p>Let’s load all the necessary libraries:</p>
<p>Also, let’s source a custom function we’ve written to visualize the distribution of cells of different datasets per cluster, alongside cluster sizes:</p>
</div>
<div id="integration-example-1-10k-pbmc-cells-profiled-using-3-and-5-10x-genomics-kits" class="section level3" number="0.1.6">
<h3><span class="header-section-number">0.1.6</span> Integration example 1: 10k PBMC cells profiled using 3’ and 5’ 10x Genomics kits</h3>
<p>Let’s load the filtered Cell Ranger <code>h5</code> matrices downloaded from 10x Genomics data repository. These can also be downloaded using the commands below.</p>
<div id="vs-5-10k-pbmc---integration-with-seurat-v3" class="section level4" number="0.1.6.1">
<h4><span class="header-section-number">0.1.6.1</span> 3’ vs 5’ 10k PBMC - Integration with Seurat v3</h4>
<p>Let’s read the data and create the appropriate <code>Seurat</code> objects. Note that 5’ dataset has other assays - namely, VDJ data.</p>
<p>Now let’s remove matrices to save memory:</p>
<p>Let’s calculate the fractions of mitochondrial genes and ribosomal proteins, and do quick-and-dirty filtering of the datasets:</p>
<p><img src="09-Integration_files/figure-html/int10-1.png" width="672" style="display: block; margin: auto;" /><img src="09-Integration_files/figure-html/int10-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now, let’s look how similar the annotations are, i.e. compare gene names. Turns out they are identical: both used the latest <code>Cell Ranger</code> annotation, GRCh38-2020A.</p>
<pre><code>## 
##  TRUE 
## 36601</code></pre>
<p>Quick filtering of the datasets removes dying cells and putative doublets:</p>
<p>Now, let’s follow <code>Seurat</code> vignette for integration. To do this we need to make a simple R list of the two objects, and normalize/find HVG for each:</p>
<p>After this, we use the two following <code>Seurat</code> commands to find integration anchors and actually perform integration. This takes about 10 min:</p>
<pre><code>## Warning in CheckDuplicateCellNames(object.list = object.list): Some cell names
## are duplicated across objects provided. Renaming to enforce unique cell names.</code></pre>
<pre><code>## Computing 2000 integration features</code></pre>
<pre><code>## Scaling features for provided objects</code></pre>
<pre><code>## Finding all pairwise anchors</code></pre>
<pre><code>## Running CCA</code></pre>
<pre><code>## Merging objects</code></pre>
<pre><code>## Finding neighborhoods</code></pre>
<pre><code>## Finding anchors</code></pre>
<pre><code>##  Found 21626 anchors</code></pre>
<pre><code>## Filtering anchors</code></pre>
<pre><code>##  Retained 8832 anchors</code></pre>
<pre><code>## Merging dataset 1 into 2</code></pre>
<pre><code>## Extracting anchors for merged samples</code></pre>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Integrating data</code></pre>
<p>Seurat integration creates a unified object that contains both original data (‘RNA’ <code>assay</code>) as well as integrated data (‘integrated’ <code>assay</code>). Let’s set the assay to RNA and visualize the datasets before integration.</p>
<p>Let’s do normalization, HVG finding, scaling, PCA, and UMAP on the un-integrated (RNA) assay:</p>
<pre><code>## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
## To use Python UMAP via reticulate, set umap.method to &#39;umap-learn&#39; and metric to &#39;correlation&#39;
## This message will be shown once per session</code></pre>
<p>UMAP plot of the datasets before integration shows clear separation. Note that we can use <code>patchwork</code> syntax with Seurat plotting functions:</p>
<p><img src="09-Integration_files/figure-html/int17-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now let’s change the assay to integrated and do the same do the same thing in the integrated assay (it’s already normalized and HVGs are selected):</p>
<p>Finally, let’s plot the integrated UMAP:</p>
<p><img src="09-Integration_files/figure-html/int19-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The data are visibly very nicely integrated. Let’s try a split plot, which should make the comparison easier:</p>
<p><img src="09-Integration_files/figure-html/int20-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now let’s cluster the integrated matrix and look how clusters are distributed between the two sets:</p>
<p><img src="09-Integration_files/figure-html/int21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We can now calculate the number of cells in each cluster that came for either 3’ or the 5’ dataset:</p>
<pre><code>##     
##      pbmc10k_3p pbmc10k_5p
##   0        1313       2140
##   1        1427        945
##   2        1214        994
##   3         898       1110
##   4         612        769
##   5         576        437
##   6         401        603
##   7         338        646
##   8         311        403
##   9         359        223
##   10        251        292
##   11        347        131
##   12        202        240
##   13        136        288
##   14        251         61
##   15        109        190
##   16        118        143
##   17         76         92
##   18         93         61
##   19         55         87
##   20         35         35
##   21         24         43
##   22         13         40
##   23         15         10
##   24         14          5
##   25          2         12</code></pre>
<p>Let’s plot the distribution among clusters using our custom function:</p>
<pre><code>## Warning: package &#39;ggplot2&#39; was built under R version 4.0.5</code></pre>
<pre><code>## Using cluster as id variables</code></pre>
<p><img src="09-Integration_files/figure-html/int23-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="vs-5-10k-pbmc---integration-with-harmony" class="section level4" number="0.1.6.2">
<h4><span class="header-section-number">0.1.6.2</span> 3’ vs 5’ 10k PBMC - Integration with Harmony</h4>
<p>Using <code>harmony</code> is much faster than pretty much any other method, and was found to perform quite well in a recent benchmark. There also are convenient wrappers for interfacing with <code>Seurat</code>. Let’s first merge the objects (without integration). Note the message about the matching cell barcodes:</p>
<pre><code>## Warning in CheckDuplicateCellNames(object.list = objects): Some cell names are
## duplicated across objects provided. Renaming to enforce unique cell names.</code></pre>
<p>Now let’s do the same as we did before:</p>
<p>Let’s plot the “before” plot again:</p>
<p><img src="09-Integration_files/figure-html/int26-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Use <code>RunHarmony</code> to run it on the combined <code>Seurat</code> object using <code>orig.ident</code> as batch:</p>
<pre><code>## Harmony 1/10</code></pre>
<pre><code>## Harmony 2/10</code></pre>
<pre><code>## Harmony 3/10</code></pre>
<pre><code>## Harmony 4/10</code></pre>
<pre><code>## Harmony 5/10</code></pre>
<pre><code>## Harmony converged after 5 iterations</code></pre>
<pre><code>## Warning: Invalid name supplied, making object name syntactically valid. New
## object name is Seurat..ProjectDim.RNA.harmony; see ?make.names for more details
## on syntax validity</code></pre>
<p><img src="09-Integration_files/figure-html/int27-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Check the generated embeddings:</p>
<pre><code>##                       harmony_1 harmony_2 harmony_3  harmony_4 harmony_5
## AAACCCACATAACTCG-1_1  -9.206607 -2.351619 -2.374652  -1.897186 -1.011885
## AAACCCACATGTAACC-1_1   7.124223 21.600131 -0.292039   1.530283 -5.792142
## AAACCCAGTGAGTCAG-1_1 -18.134725  3.405369  5.256459   4.220001  3.961466
## AAACGAACAGTCAGTT-1_1 -18.103262 15.279955 12.301681 -18.115094 31.785955
## AAACGAACATTCGGGC-1_1  11.097966 -2.330278 -2.723953   1.546468  1.552332</code></pre>
<p>Check the PCA plot after:</p>
<p><img src="09-Integration_files/figure-html/int29-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Do UMAP and clustering:</p>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 19190
## Number of edges: 280811
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8952
## Number of communities: 26
## Elapsed time: 1 seconds</code></pre>
<p>Finally, so same UMAP plots of integrated datasets as above:</p>
<p><img src="09-Integration_files/figure-html/int31-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><img src="09-Integration_files/figure-html/int32-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>These look a bit worse than ones from <code>Seurat</code>:</p>
<p><img src="09-Integration_files/figure-html/int33-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Finally, let’s take a look at the cluster content:</p>
<pre><code>## Using cluster as id variables</code></pre>
<p><img src="09-Integration_files/figure-html/int34-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Clusters and their content look pretty similar to what we obtained after <code>Seurat</code> integration. For a more detailed analysis, we would need cell type assignments.</p>
</div>
<div id="vs-5-10k-pbmc---integration-with-liger" class="section level4" number="0.1.6.3">
<h4><span class="header-section-number">0.1.6.3</span> 3’ vs 5’ 10k PBMC - Integration with LIGER</h4>
<p>Similarly to other methods, we make a unified object and normalize/HVG/scale it. LIGER does not center data when scaling, hence the <code>do.center</code> option in <code>ScaleData</code>). The last two functions are wrappers that run <code>rliger</code> using <code>orig.ident</code> as a batch variable.</p>
<pre><code>## Warning in CheckDuplicateCellNames(object.list = objects): Some cell names are
## duplicated across objects provided. Renaming to enforce unique cell names.</code></pre>
<pre><code>## Scaling data matrix</code></pre>
<pre><code>## Scaling data from split pbmc10k_3p</code></pre>
<pre><code>## Scaling data from split pbmc10k_5p</code></pre>
<pre><code>## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   3%
  |                                                                            
  |=====                                                                 |   7%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=========                                                             |  13%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |==========================                                            |  37%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |==============================                                        |  43%
  |                                                                            
  |=================================                                     |  47%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |=====================================                                 |  53%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |============================================                          |  63%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |=============================================================         |  87%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |=================================================================     |  93%
  |                                                                            
  |====================================================================  |  97%
  |                                                                            
  |======================================================================| 100%
## Finished in 10.19865 mins, 30 iterations.
## Max iterations set: 30.
## Final objective delta: 1.924923e-05.
## Best results with seed 1.</code></pre>
<pre><code>## Warning: No columnames present in cell embeddings, setting to &#39;riNMF_1:30&#39;</code></pre>
<p>You can optionally perform Louvain clustering (<code>FindNeighbors</code> and <code>FindClusters</code>) after <code>RunQuantileNorm</code> - we’ll do this as well to compare the results to the previous integration approaches. We use the same parameters (k = 10 for neighbors, default resolution for Louvain clustering).</p>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 19190
## Number of edges: 281100
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.8986
## Number of communities: 29
## Elapsed time: 1 seconds</code></pre>
<p>As previously, let’s do dimensionality reduction and plotting:</p>
<p><img src="09-Integration_files/figure-html/int37-1.png" width="672" style="display: block; margin: auto;" /><img src="09-Integration_files/figure-html/int37-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Clustering appears to be somewhat finer with the <code>LIGER</code>-integrated data:</p>
<p><img src="09-Integration_files/figure-html/int38-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The clusters look pretty different, and per-cluster distribution seems to confirm this (two clusters were deemed unique to 3’ and 5’ dataset, respectively):</p>
<pre><code>## Using cluster as id variables</code></pre>
<p><img src="09-Integration_files/figure-html/int39-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="dataset-2-10k-pbmc-cells-3-10x-and-sorted-whole-blood-generated-using-strt-seq" class="section level3" number="0.1.7">
<h3><span class="header-section-number">0.1.7</span> DATASET 2: 10k PBMC cells (3’ 10x) and sorted whole blood generated using STRT-Seq</h3>
<p>Although we already have all the necessary files in our <code>/data</code> folder, we can download the necessary files from GEO database:</p>
<p>Next, let’s make <code>Seurat</code> objects and re-define some of the metadata columns (GEO dataset simply puts the cell type into the <code>orig.ident</code> slot, which will interfere with what we want to do next):</p>
<pre><code>## Warning: Non-unique features (rownames) present in the input matrix, making
## unique</code></pre>
<pre><code>##                    cell_type nCount_RNA nFeature_RNA  orig.ident
## BNK_spBM1_L1_bar25       BNK      24494         1869 whole_blood
## BNK_spBM1_L1_bar26       BNK      61980         2051 whole_blood
## BNK_spBM1_L1_bar27       BNK     124382         3872 whole_blood
## BNK_spBM1_L1_bar28       BNK       8144         1475 whole_blood
## BNK_spBM1_L1_bar29       BNK      53612         2086 whole_blood
## BNK_spBM1_L1_bar30       BNK      33582         2038 whole_blood</code></pre>
<p>Do basic quality control. STRT-Seq is quite different from 10x and has a lot more detected genes per cell. Also, for some reason we don’t have the MT genes in the quantified matrix of the whole blood dataset. That’s unfortunate, but not critical.</p>
<p><img src="09-Integration_files/figure-html/int44-1.png" width="672" style="display: block; margin: auto;" /><img src="09-Integration_files/figure-html/int44-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>The annotation that was used to process the GEO whole blood dataset is quite different from the Cell Ranger GRCh38-2020A. Let’s see how many common genes are there:</p>
<pre><code>## 
## FALSE  TRUE 
## 18050 18551</code></pre>
<p>Let’s filter the cells with too high or too low number of genes, or too high MT gene content. Also, let’s limit the individual matrices to common genes only:</p>
<div id="k-pbmc-3-10x-with-strt-seq-whole-blood---integration-with-seurat" class="section level4" number="0.1.7.1">
<h4><span class="header-section-number">0.1.7.1</span> 10k PBMC 3’ 10x with STRT-Seq whole blood - integration with Seurat</h4>
<p>As previously, let’s make a list and normalize/find HVG for each object:</p>
<p>Here we actually do the integration. Seurat 3 does it in two steps.</p>
<pre><code>## Computing 2000 integration features</code></pre>
<pre><code>## Scaling features for provided objects</code></pre>
<pre><code>## Finding all pairwise anchors</code></pre>
<pre><code>## Running CCA</code></pre>
<pre><code>## Merging objects</code></pre>
<pre><code>## Finding neighborhoods</code></pre>
<pre><code>## Finding anchors</code></pre>
<pre><code>##  Found 11353 anchors</code></pre>
<pre><code>## Filtering anchors</code></pre>
<pre><code>##  Retained 3276 anchors</code></pre>
<pre><code>## Merging dataset 2 into 1</code></pre>
<pre><code>## Extracting anchors for merged samples</code></pre>
<pre><code>## Finding integration vectors</code></pre>
<pre><code>## Finding integration vector weights</code></pre>
<pre><code>## Integrating data</code></pre>
<p>Let’s do the basic processing and visualization of the uncorrected dataset:</p>
<p><img src="09-Integration_files/figure-html/int49-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now, let’s take a look at the integrated dataset (it’s already normalized and HVGs are selected):</p>
<p><img src="09-Integration_files/figure-html/int50-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Let’s look at some markers:</p>
<p><img src="09-Integration_files/figure-html/int51-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>From the plot we can see that there are some significant cell types that are absent from PBMC dataset, but exist in the whole blood dataset. LTF gene is the most prominent marker of neutrophils, and HBA1 is a hemoglobin gene expressed in erythrocytes.</p>
<p>Now let’s cluster the integrated matrix and look how clusters are distributed between the two sets:</p>
<p><img src="09-Integration_files/figure-html/int52-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Cluster composition shows many clusters unique to the whole blood dataset:</p>
<pre><code>##     
##      pbmc10k_3p whole_blood
##   0        1426         237
##   1        1385          71
##   2        1264         130
##   3        1211         112
##   4        1115         145
##   5           0        1052
##   6         355         467
##   7         377         211
##   8         386         199
##   9           0         550
##   10        343         157
##   11        390          82
##   12          0         441
##   13        283         125
##   14          7         388
##   15          3         380
##   16         19         362
##   17          4         367
##   18          2         316
##   19        297          13
##   20          0         308
##   21          0         265
##   22          0         222
##   23          0         221
##   24         15         179
##   25        106          22
##   26         93          19
##   27          0         103
##   28         77           3
##   29          0          50
##   30         32           2
##   31          0          32</code></pre>
<pre><code>## Using cluster as id variables</code></pre>
<p><img src="09-Integration_files/figure-html/int54-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>We can take advantage of the metadata that was present in GSE149938:</p>
<pre><code>## 
##    BNK    CMP    ery   immB    MEP    MLP   preB   proB toxiNK 
##      1      2   1040      3      1      1      2      1      1</code></pre>
<pre><code>## 
##     BNK     CLP     CMP     HSC    immB  kineNK matureN   metaN     MLP    myeN 
##       2       2       1       1       1      55       1       7       1      99 
##  plasma    proN  toxiNK 
##       1     136       1</code></pre>
<pre><code>## 
##    ery   naiB plasma 
##     11      2    166</code></pre>
<pre><code>## 
##     BNK     CLP     CMP     ery     HSC    LMPP matureN     MEP     MPP     NKP 
##       1       3      61       4      72       1       1     144      74       1</code></pre>
</div>
<div id="k-pbmc-3-10x-with-strt-seq-whole-blood---integration-with-harmony" class="section level4" number="0.1.7.2">
<h4><span class="header-section-number">0.1.7.2</span> 10k PBMC 3’ 10x with STRT-Seq whole blood - integration with Harmony</h4>
<p>Similarly to the previous approaches, let’s make a merged <code>Seurat</code> dataset, normalize and process it:</p>
<p>We can take a look at the PCA plot for a change, as well as distributions along the first principal component:</p>
<p><img src="09-Integration_files/figure-html/int57-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>UMAP also shows clear differences between the datasets:</p>
<p><img src="09-Integration_files/figure-html/int58-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Let’s run <code>harmony</code> using a simple wrapper named <code>RunHarmony</code> from <code>SeuratWrappers</code> library:</p>
<pre><code>## Harmony 1/10</code></pre>
<pre><code>## Harmony 2/10</code></pre>
<pre><code>## Harmony 3/10</code></pre>
<pre><code>## Harmony converged after 3 iterations</code></pre>
<pre><code>## Warning: Invalid name supplied, making object name syntactically valid. New
## object name is Seurat..ProjectDim.RNA.harmony; see ?make.names for more details
## on syntax validity</code></pre>
<p><img src="09-Integration_files/figure-html/int59-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>This generates the embeddings that we shall later use for all downstream analysis.</p>
<pre><code>##                    harmony_1  harmony_2 harmony_3   harmony_4  harmony_5
## AAACCCACATAACTCG-1  2.592137 -1.6045869  3.192993 -0.03594751  3.8447941
## AAACCCACATGTAACC-1  4.244764  3.2122684 -9.738222 -5.90380632  0.9607984
## AAACCCAGTGAGTCAG-1  3.208084 -2.1054920  2.035577  1.90984384  5.3665634
## AAACGAACAGTCAGTT-1 -1.106694  1.8151680  3.092745 -2.34038488  7.6785360
## AAACGAACATTCGGGC-1  4.735928 -0.4421468 -2.196355  2.77622970 -3.2385050</code></pre>
<p>Corrected PCA and distribution:</p>
<p><img src="09-Integration_files/figure-html/int61-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Run UMAP and perform Louvain clustering:</p>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 16421
## Number of edges: 266804
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.9230
## Number of communities: 31
## Elapsed time: 1 seconds</code></pre>
<pre><code>## 1 singletons identified. 30 final clusters.</code></pre>
<p><img src="09-Integration_files/figure-html/int63-1.png" width="672" style="display: block; margin: auto;" /><img src="09-Integration_files/figure-html/int63-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Corrected results for this dataset appear to be very similar to Seurat v3:</p>
<p><img src="09-Integration_files/figure-html/int64-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>More detailed cluster examination also seems to confirm this:</p>
<pre><code>## Using cluster as id variables</code></pre>
<p><img src="09-Integration_files/figure-html/int65-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="k-pbmc-3-10x-with-strt-seq-whole-blood---integration-with-liger" class="section level4" number="0.1.7.3">
<h4><span class="header-section-number">0.1.7.3</span> 10k PBMC 3’ 10x with STRT-Seq whole blood - integration with LIGER</h4>
<p>Finally, let’s do data integration with <code>LIGER</code>. This step takes several minutes to run:</p>
<pre><code>## Scaling data matrix</code></pre>
<pre><code>## Scaling data from split pbmc10k_3p</code></pre>
<pre><code>## Scaling data from split whole_blood</code></pre>
<pre><code>## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   3%
  |                                                                            
  |=====                                                                 |   7%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=========                                                             |  13%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |==========================                                            |  37%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |==============================                                        |  43%
  |                                                                            
  |=================================                                     |  47%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |=====================================                                 |  53%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |============================================                          |  63%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |=============================================================         |  87%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |=================================================================     |  93%
  |                                                                            
  |====================================================================  |  97%
  |                                                                            
  |======================================================================| 100%
## Finished in 8.824175 mins, 30 iterations.
## Max iterations set: 30.
## Final objective delta: 7.198128e-05.
## Best results with seed 1.</code></pre>
<pre><code>## Warning: No columnames present in cell embeddings, setting to &#39;riNMF_1:30&#39;</code></pre>
<p>We will then perform Louvain clustering (<code>FindNeighbors</code> and <code>FindClusters</code>) with the settings similar to what we have been using before:</p>
<pre><code>## Computing nearest neighbor graph</code></pre>
<pre><code>## Computing SNN</code></pre>
<pre><code>## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck
## 
## Number of nodes: 16421
## Number of edges: 267813
## 
## Running Louvain algorithm...
## Maximum modularity in 10 random starts: 0.9197
## Number of communities: 35
## Elapsed time: 0 seconds</code></pre>
<p>Let’s look at the corrected UMAP plot in a couple of different ways:</p>
<p><img src="09-Integration_files/figure-html/int68-1.png" width="672" style="display: block; margin: auto;" /><img src="09-Integration_files/figure-html/int68-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>Finally, a look at distribution of datasets per cluster:</p>
<pre><code>## Using cluster as id variables</code></pre>
<p><img src="09-Integration_files/figure-html/int69-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="sessioninfo" class="section level3" number="0.1.8">
<h3><span class="header-section-number">0.1.8</span> sessionInfo()</h3>
<pre><code>## R version 4.0.4 (2021-02-15)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ggplot2_3.3.4          org.Hs.eg.db_3.12.0    AnnotationDbi_1.52.0  
##  [4] IRanges_2.24.1         S4Vectors_0.28.1       Biobase_2.50.0        
##  [7] BiocGenerics_0.36.0    dplyr_1.0.5            RColorBrewer_1.1-2    
## [10] reshape2_1.4.4         rliger_1.0.0           Matrix_1.3-2          
## [13] cowplot_1.1.1          harmony_1.0            Rcpp_1.0.6            
## [16] patchwork_1.1.1        SeuratWrappers_0.3.0   panc8.SeuratData_3.0.2
## [19] SeuratData_0.2.1       SeuratDisk_0.0.0.9019  SeuratObject_4.0.0    
## [22] Seurat_4.0.1           knitr_1.31            
## 
## loaded via a namespace (and not attached):
##   [1] plyr_1.8.6            igraph_1.2.6          lazyeval_0.2.2       
##   [4] splines_4.0.4         listenv_0.8.0         scattermore_0.7      
##   [7] digest_0.6.27         foreach_1.5.1         htmltools_0.5.1.1    
##  [10] fansi_0.4.2           memoise_2.0.0         magrittr_2.0.1       
##  [13] tensor_1.5            cluster_2.1.1         doParallel_1.0.16    
##  [16] ROCR_1.0-11           remotes_2.2.0         globals_0.14.0       
##  [19] matrixStats_0.58.0    R.utils_2.10.1        spatstat.sparse_2.0-0
##  [22] colorspace_2.0-0      blob_1.2.1            rappdirs_0.3.3       
##  [25] ggrepel_0.9.1         xfun_0.22             crayon_1.4.1         
##  [28] jsonlite_1.7.2        spatstat.data_2.1-0   survival_3.2-10      
##  [31] zoo_1.8-9             iterators_1.0.13      glue_1.4.2           
##  [34] polyclip_1.10-0       gtable_0.3.0          leiden_0.3.7         
##  [37] future.apply_1.7.0    abind_1.4-5           scales_1.1.1         
##  [40] DBI_1.1.1             miniUI_0.1.1.1        riverplot_0.10       
##  [43] viridisLite_0.3.0     xtable_1.8-4          reticulate_1.20      
##  [46] spatstat.core_2.0-0   mclust_5.4.7          bit_4.0.4            
##  [49] rsvd_1.0.3            htmlwidgets_1.5.3     httr_1.4.2           
##  [52] FNN_1.1.3             ellipsis_0.3.1        ica_1.0-2            
##  [55] farver_2.1.0          pkgconfig_2.0.3       R.methodsS3_1.8.1    
##  [58] sass_0.3.1            uwot_0.1.10           deldir_0.2-10        
##  [61] utf8_1.2.1            labeling_0.4.2        tidyselect_1.1.0     
##  [64] rlang_0.4.10          later_1.1.0.1         cachem_1.0.4         
##  [67] munsell_0.5.0         tools_4.0.4           cli_2.3.1            
##  [70] RSQLite_2.2.4         generics_0.1.0        ggridges_0.5.3       
##  [73] evaluate_0.14         stringr_1.4.0         fastmap_1.1.0        
##  [76] yaml_2.2.1            goftest_1.2-2         bit64_4.0.5          
##  [79] fitdistrplus_1.1-3    purrr_0.3.4           RANN_2.6.1           
##  [82] pbapply_1.4-3         future_1.21.0         nlme_3.1-152         
##  [85] mime_0.10             R.oo_1.24.0           hdf5r_1.3.3          
##  [88] compiler_4.0.4        rstudioapi_0.13       plotly_4.9.3         
##  [91] png_0.1-7             spatstat.utils_2.1-0  tibble_3.1.0         
##  [94] bslib_0.2.4           stringi_1.5.3         highr_0.8            
##  [97] RSpectra_0.16-0       lattice_0.20-41       vctrs_0.3.6          
## [100] pillar_1.5.1          lifecycle_1.0.0       BiocManager_1.30.10  
## [103] spatstat.geom_2.0-1   lmtest_0.9-38         jquerylib_0.1.3      
## [106] RcppAnnoy_0.0.18      data.table_1.14.0     irlba_2.3.3          
## [109] httpuv_1.5.5          R6_2.5.0              bookdown_0.22        
## [112] promises_1.2.0.1      KernSmooth_2.23-18    gridExtra_2.3        
## [115] parallelly_1.24.0     codetools_0.2-18      MASS_7.3-53.1        
## [118] assertthat_0.2.1      withr_2.4.1           sctransform_0.3.2    
## [121] mgcv_1.8-34           grid_4.0.4            rpart_4.1-15         
## [124] tidyr_1.1.3           rmarkdown_2.7         Rtsne_0.15           
## [127] shiny_1.6.0</code></pre>
</div>
</div>
            </section>

          </div>
        </div>
      </div>


    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"search": false
});
});
</script>

</body>

</html>
