<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>1 Basic Quality Control (QC) and Exploration of scRNA-seq Datasets | 06a-QC_and_confounders.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="1 Basic Quality Control (QC) and Exploration of scRNA-seq Datasets | 06a-QC_and_confounders.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="1 Basic Quality Control (QC) and Exploration of scRNA-seq Datasets | 06a-QC_and_confounders.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  


<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<!-- for Facebook -->  
<meta property="og:url" content="http://hemberg-lab.github.io/scRNA.seq.course/" />
<meta property="og:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="Analysis of single-cell RNA-seq data" />
<meta name="twitter:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-71525309-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-71525309-1');
</script>

<!-- For Logo -->
<script>
  $(document).ready(function(){
    $(".book-header").prepend('<img id="SangerLogo" src="figures/sanger-logo-header.png" />')
  }
</script>

<script>
  $(document).ready(function(){
    $('.book-header').wrap( $('<a>').attr('href', 'https://www.sanger.ac.uk/').attr('target', '_blank'))
  });
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path=""><a href="#basic-quality-control-qc-and-exploration-of-scrna-seq-datasets"><i class="fa fa-check"></i><b>1</b> Basic Quality Control (QC) and Exploration of scRNA-seq Datasets</a>
<ul>
<li class="chapter" data-level="1.1" data-path=""><a href="#dataset-contruction-and-qc"><i class="fa fa-check"></i><b>1.1</b> Dataset Contruction and QC</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path=""><a href="#introduction"><i class="fa fa-check"></i><b>1.1.1</b> Introduction</a></li>
<li class="chapter" data-level="1.1.2" data-path=""><a href="#tung-dataset"><i class="fa fa-check"></i><b>1.1.2</b> Tung Dataset</a></li>
<li class="chapter" data-level="1.1.3" data-path=""><a href="#basic-qc"><i class="fa fa-check"></i><b>1.1.3</b> Basic QC</a></li>
<li class="chapter" data-level="1.1.4" data-path=""><a href="#highly-expressed-genes"><i class="fa fa-check"></i><b>1.1.4</b> Highly Expressed Genes</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path=""><a href="#data-visualization-and-dimensionality-reduction"><i class="fa fa-check"></i><b>1.2</b> Data Visualization and Dimensionality Reduction</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path=""><a href="#introduction-1"><i class="fa fa-check"></i><b>1.2.1</b> Introduction</a></li>
<li class="chapter" data-level="1.2.2" data-path=""><a href="#pca-plot"><i class="fa fa-check"></i><b>1.2.2</b> PCA plot</a></li>
<li class="chapter" data-level="1.2.3" data-path=""><a href="#tsne-map"><i class="fa fa-check"></i><b>1.2.3</b> tSNE Map</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path=""><a href="#identifying-confounding-factors"><i class="fa fa-check"></i><b>1.3</b> Identifying Confounding Factors</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path=""><a href="#introduction-2"><i class="fa fa-check"></i><b>1.3.1</b> Introduction</a></li>
<li class="chapter" data-level="1.3.2" data-path=""><a href="#correlations-with-pcs"><i class="fa fa-check"></i><b>1.3.2</b> Correlations with PCs</a></li>
<li class="chapter" data-level="1.3.3" data-path=""><a href="#explanatory-variables"><i class="fa fa-check"></i><b>1.3.3</b> Explanatory Variables</a></li>
<li class="chapter" data-level="1.3.4" data-path=""><a href="#other-confounders"><i class="fa fa-check"></i><b>1.3.4</b> Other Confounders</a></li>
<li class="chapter" data-level="1.3.5" data-path=""><a href="#exercise"><i class="fa fa-check"></i><b>1.3.5</b> Exercise</a></li>
<li class="chapter" data-level="1.3.6" data-path=""><a href="#sessioninfo"><i class="fa fa-check"></i><b>1.3.6</b> sessionInfo()</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:title:end-->
<!--bookdown:title:start-->
<div id="basic-quality-control-qc-and-exploration-of-scrna-seq-datasets" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Basic Quality Control (QC) and Exploration of scRNA-seq Datasets</h1>
<div id="dataset-contruction-and-qc" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Dataset Contruction and QC</h2>
<div id="introduction" class="section level3" number="1.1.1">
<h3><span class="header-section-number">1.1.1</span> Introduction</h3>
<p>Once gene expression has been quantified it is summarized as an <strong>expression matrix</strong> where each row corresponds to a gene (or transcript) and each column corresponds to a single cell. In the next step, the matrix should be examined to remove poor quality cells. Failure to remove low quality cells at this stage may add technical noise which has the potential to obscure the biological signals of interest in the downstream analysis.</p>
<p>Since there is currently no standard method for performing scRNA-seq, the expected values for the various QC measures that will be presented here can vary substantially from experiment to experiment. Thus, to perform QC we will be looking for cells which are outliers with respect to the rest of the dataset rather than comparing to independent quality standards. Consequently, care should be taken when comparing quality metrics across datasets sequenced using different protocols.</p>
</div>
<div id="tung-dataset" class="section level3" number="1.1.2">
<h3><span class="header-section-number">1.1.2</span> Tung Dataset</h3>
<p>To illustrate cell QC, we consider a <a href="http://jdblischak.github.io/singleCellSeq/analysis/">dataset</a> of induced pluripotent stem cells generated from three different individuals <span class="citation">[@Tung2017-ba]</span> in <a href="http://giladlab.uchicago.edu/">Yoav Gilad</a>’s lab at the University of Chicago. The experiments were carried out on the Fluidigm C1 platform and to facilitate the quantification both unique molecular identifiers (UMIs) and ERCC <em>spike-ins</em> were used. Due to rapid increase in droplet-based method use, spike-ins are not widely used anymore; however, they can serve as an informative control for low throughput methods. The data files are located in the <code>tung</code> folder in your working directory. These files are the copies of the original files made on the 15/03/16. We will use these copies for reproducibility purposes.</p>
<p>We’ll use <code>scater</code> package, as well as <code>AnnotationDbi</code> and <code>org.Hs.eg.db</code> to convert ENSEMBL IDs into gene names (symbols).</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(AnnotationDbi)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnsDb.Hsapiens.v86)</span></code></pre></div>
<p>Next we’ll read in the matrix and the per-cell annotation. The latter is converted to factors on the fly.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>molecules <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;data/tung/molecules.txt&quot;</span>,<span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>annotation <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;data/tung/annotation.txt&quot;</span>,<span class="at">stringsAsFactors =</span> T)</span></code></pre></div>
<p>Take a quick look at the dataset:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(molecules[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code></pre></div>
<pre><code>##                 NA19098.r1.A01 NA19098.r1.A02 NA19098.r1.A03
## ENSG00000237683              0              0              0
## ENSG00000187634              0              0              0
## ENSG00000188976              3              6              1
## ENSG00000187961              0              0              0
## ENSG00000187583              0              0              0
## ENSG00000187642              0              0              0</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(annotation)</span></code></pre></div>
<pre><code>##   individual replicate well      batch      sample_id
## 1    NA19098        r1  A01 NA19098.r1 NA19098.r1.A01
## 2    NA19098        r1  A02 NA19098.r1 NA19098.r1.A02
## 3    NA19098        r1  A03 NA19098.r1 NA19098.r1.A03
## 4    NA19098        r1  A04 NA19098.r1 NA19098.r1.A04
## 5    NA19098        r1  A05 NA19098.r1 NA19098.r1.A05
## 6    NA19098        r1  A06 NA19098.r1 NA19098.r1.A06</code></pre>
<p>Here we set <code>altExp</code> to contain ERCC, removing ERCC features from the main object:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">&lt;-</span> <span class="fu">SingleCellExperiment</span>(<span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> <span class="fu">as.matrix</span>(molecules)), <span class="at">colData =</span> annotation)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">altExp</span>(umi,<span class="st">&quot;ERCC&quot;</span>) <span class="ot">&lt;-</span> umi[<span class="fu">grep</span>(<span class="st">&quot;^ERCC-&quot;</span>,<span class="fu">rownames</span>(umi)), ]</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">&lt;-</span> umi[<span class="fu">grep</span>(<span class="st">&quot;^ERCC-&quot;</span>,<span class="fu">rownames</span>(umi),<span class="at">invert =</span> T), ]</span></code></pre></div>
<p>Now, let’s map ENSEMBL IDs to gene symbols. From the <code>table</code> command, we can see that most genes were annotated; however, 846 returned “NA”. By default, <code>mapIds</code> returs one symbol per ID; this behaviour can be changed using <code>multiVals</code> argument.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>gene_names <span class="ot">&lt;-</span> <span class="fu">mapIds</span>(org.Hs.eg.db, <span class="at">keys=</span><span class="fu">rownames</span>(umi), <span class="at">keytype=</span><span class="st">&quot;ENSEMBL&quot;</span>, <span class="at">columns=</span><span class="st">&quot;SYMBOL&quot;</span>,<span class="at">column=</span><span class="st">&quot;SYMBOL&quot;</span>)</span></code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(umi)<span class="sc">$</span>SYMBOL <span class="ot">&lt;-</span> gene_names</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">is.na</span>(gene_names))</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
## 18092   846</code></pre>
<p>Let’s remove all genes for which no symbols were found:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">&lt;-</span> umi[<span class="sc">!</span> <span class="fu">is.na</span>(<span class="fu">rowData</span>(umi)<span class="sc">$</span>SYMBOL),]</span></code></pre></div>
<p>Let’s check if we can find mitochondrial proteins in the newly annotated symbols.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">&quot;^MT-&quot;</span>,<span class="fu">rowData</span>(umi)<span class="sc">$</span>SYMBOL,<span class="at">value =</span> T)</span></code></pre></div>
<pre><code>## named character(0)</code></pre>
<p>Strangely, this returns nothing. Similar command to find ribosomal proteins (which start with RPL or RPS) works as expected:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">&quot;^RP[LS]&quot;</span>,<span class="fu">rowData</span>(umi)<span class="sc">$</span>SYMBOL,<span class="at">value =</span> T)</span></code></pre></div>
<pre><code>## ENSG00000116251 ENSG00000142676 ENSG00000117676 ENSG00000142937 ENSG00000122406 
##         &quot;RPL22&quot;         &quot;RPL11&quot;       &quot;RPS6KA1&quot;          &quot;RPS8&quot;          &quot;RPL5&quot; 
## ENSG00000177954 ENSG00000136643 ENSG00000138326 ENSG00000177600 ENSG00000166441 
##         &quot;RPS27&quot;       &quot;RPS6KC1&quot;         &quot;RPS24&quot;         &quot;RPLP2&quot;        &quot;RPL27A&quot; 
## ENSG00000110700 ENSG00000162302 ENSG00000175634 ENSG00000149273 ENSG00000118181 
##         &quot;RPS13&quot;       &quot;RPS6KA4&quot;       &quot;RPS6KB2&quot;          &quot;RPS3&quot;         &quot;RPS25&quot; 
## ENSG00000197728 ENSG00000229117 ENSG00000089009 ENSG00000089157 ENSG00000122026 
##         &quot;RPS26&quot;         &quot;RPL41&quot;          &quot;RPL6&quot;         &quot;RPLP0&quot;         &quot;RPL21&quot; 
## ENSG00000165496 ENSG00000213741 ENSG00000165502 ENSG00000198208 ENSG00000100784 
##        &quot;RPL10L&quot;         &quot;RPS29&quot;       &quot;RPL36AL&quot;       &quot;RPS6KL1&quot;       &quot;RPS6KA5&quot; 
## ENSG00000185088 ENSG00000174444 ENSG00000137818 ENSG00000182774 ENSG00000140986 
##        &quot;RPS27L&quot;          &quot;RPL4&quot;         &quot;RPLP1&quot;         &quot;RPS17&quot;         &quot;RPL3L&quot; 
## ENSG00000140988 ENSG00000134419 ENSG00000167526 ENSG00000161970 ENSG00000198242 
##          &quot;RPS2&quot;        &quot;RPS15A&quot;         &quot;RPL13&quot;         &quot;RPL26&quot;        &quot;RPL23A&quot; 
## ENSG00000125691 ENSG00000108298 ENSG00000131469 ENSG00000108443 ENSG00000172809 
##         &quot;RPL23&quot;         &quot;RPL19&quot;         &quot;RPL27&quot;       &quot;RPS6KB1&quot;         &quot;RPL38&quot; 
## ENSG00000265681 ENSG00000115268 ENSG00000130255 ENSG00000233927 ENSG00000105640 
##         &quot;RPL17&quot;         &quot;RPS15&quot;         &quot;RPL36&quot;         &quot;RPS28&quot;        &quot;RPL18A&quot; 
## ENSG00000105193 ENSG00000105372 ENSG00000063177 ENSG00000142541 ENSG00000142534 
##         &quot;RPS16&quot;         &quot;RPS19&quot;         &quot;RPL18&quot;        &quot;RPL13A&quot;         &quot;RPS11&quot; 
## ENSG00000170889 ENSG00000108107 ENSG00000083845 ENSG00000171863 ENSG00000143947 
##          &quot;RPS9&quot;         &quot;RPL28&quot;          &quot;RPS5&quot;          &quot;RPS7&quot;        &quot;RPS27A&quot; 
## ENSG00000071082 ENSG00000197756 ENSG00000171858 ENSG00000100316 ENSG00000187051 
##         &quot;RPL31&quot;        &quot;RPL37A&quot;         &quot;RPS21&quot;          &quot;RPL3&quot;      &quot;RPS19BP1&quot; 
## ENSG00000144713 ENSG00000174748 ENSG00000168028 ENSG00000188846 ENSG00000162244 
##         &quot;RPL32&quot;         &quot;RPL15&quot;          &quot;RPSA&quot;         &quot;RPL14&quot;         &quot;RPL29&quot; 
## ENSG00000114391 ENSG00000163584 ENSG00000163923 ENSG00000182899 ENSG00000163682 
##         &quot;RPL24&quot;       &quot;RPL22L1&quot;        &quot;RPL39L&quot;        &quot;RPL35A&quot;          &quot;RPL9&quot; 
## ENSG00000109475 ENSG00000145425 ENSG00000145592 ENSG00000186468 ENSG00000164587 
##         &quot;RPL34&quot;         &quot;RPS3A&quot;         &quot;RPL37&quot;         &quot;RPS23&quot;         &quot;RPS14&quot; 
## ENSG00000037241 ENSG00000231500 ENSG00000124614 ENSG00000198755 ENSG00000146223 
##       &quot;RPL26L1&quot;         &quot;RPS18&quot;         &quot;RPS10&quot;        &quot;RPL10A&quot;        &quot;RPL7L1&quot; 
## ENSG00000112306 ENSG00000071242 ENSG00000008988 ENSG00000147604 ENSG00000156482 
##         &quot;RPS12&quot;       &quot;RPS6KA2&quot;         &quot;RPS20&quot;          &quot;RPL7&quot;         &quot;RPL30&quot; 
## ENSG00000161016 ENSG00000137154 ENSG00000136942 ENSG00000197958 ENSG00000148303 
##          &quot;RPL8&quot;          &quot;RPS6&quot;         &quot;RPL35&quot;         &quot;RPL12&quot;         &quot;RPL7A&quot; 
## ENSG00000177189 ENSG00000198034 ENSG00000072133 ENSG00000241343 ENSG00000198918 
##       &quot;RPS6KA3&quot;         &quot;RPS4X&quot;       &quot;RPS6KA6&quot;        &quot;RPL36A&quot;         &quot;RPL39&quot; 
## ENSG00000147403 ENSG00000129824 
##         &quot;RPL10&quot;        &quot;RPS4Y1&quot;</code></pre>
<p>Quick search for mitochondrial protein <em>ATP8</em>, which is also called <em>MT-ATP8</em>, shows that the name does not contain “MT-”. However, the correct feature (ENSEMBL ID <em>ENSG00000228253</em>) is present in our annotation.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span>(<span class="st">&quot;ATP8&quot;</span>,<span class="fu">rowData</span>(umi)<span class="sc">$</span>SYMBOL,<span class="at">value =</span> T)</span></code></pre></div>
<pre><code>## ENSG00000143515 ENSG00000132932 ENSG00000104043 ENSG00000081923 ENSG00000130270 
##        &quot;ATP8B2&quot;        &quot;ATP8A2&quot;        &quot;ATP8B4&quot;        &quot;ATP8B1&quot;        &quot;ATP8B3&quot; 
## ENSG00000124406 ENSG00000228253 
##        &quot;ATP8A1&quot;          &quot;ATP8&quot;</code></pre>
<p>Most modern annotations, e.g. ones used by <code>Cell Ranger</code>, will have mitochondrial genes names that start with <em>MT-</em>. For some reason, the one we have found does not. Annotation problems in general are very common and should be always considered carefully. In our case, we also can’t find the location of genes since chromosomes are not supported in <code>org.Hs.eg.db</code> - there are no genome location columns in this database:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">columns</span>(org.Hs.eg.db)</span></code></pre></div>
<pre><code>##  [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot;
##  [6] &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;     &quot;EVIDENCEALL&quot;  &quot;GENENAME&quot;    
## [11] &quot;GO&quot;           &quot;GOALL&quot;        &quot;IPI&quot;          &quot;MAP&quot;          &quot;OMIM&quot;        
## [16] &quot;ONTOLOGY&quot;     &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PFAM&quot;         &quot;PMID&quot;        
## [21] &quot;PROSITE&quot;      &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UCSCKG&quot;       &quot;UNIGENE&quot;     
## [26] &quot;UNIPROT&quot;</code></pre>
<p>Let’s try a different, more detailed database - <code>EnsDb.Hsapiens.v86</code>. Using this resource, we can find 13 protein-coding genes located in the mitochondrion:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ensdb_genes <span class="ot">&lt;-</span> <span class="fu">genes</span>(EnsDb.Hsapiens.v86)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>MT_names <span class="ot">&lt;-</span> ensdb_genes[<span class="fu">seqnames</span>(ensdb_genes) <span class="sc">==</span> <span class="st">&quot;MT&quot;</span>]<span class="sc">$</span>gene_id</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>is_mito <span class="ot">&lt;-</span> <span class="fu">rownames</span>(umi) <span class="sc">%in%</span> MT_names</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(is_mito)</span></code></pre></div>
<pre><code>## is_mito
## FALSE  TRUE 
## 18079    13</code></pre>
</div>
<div id="basic-qc" class="section level3" number="1.1.3">
<h3><span class="header-section-number">1.1.3</span> Basic QC</h3>
<p>The following <code>scater</code> functions allow us to add per-cell and per-gene metrics useful for dataset evaluation. Most popular metrics per cell are total number of counts (UMIs), total number of detected genes, total number of mitochondrial counts, percent of mitochondrial counts, etc.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>umi_cell <span class="ot">&lt;-</span> <span class="fu">perCellQCMetrics</span>(umi,<span class="at">subsets=</span><span class="fu">list</span>(<span class="at">Mito=</span>is_mito))</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>umi_feature <span class="ot">&lt;-</span> <span class="fu">perFeatureQCMetrics</span>(umi)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(umi_cell)</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 9 columns
##                      sum  detected subsets_Mito_sum subsets_Mito_detected
##                &lt;numeric&gt; &lt;numeric&gt;        &lt;numeric&gt;             &lt;numeric&gt;
## NA19098.r1.A01     61712      8246             4883                    13
## NA19098.r1.A02     62301      8114             3732                    13
## NA19098.r1.A03     42213      7190             3089                    13
## NA19098.r1.A04     52329      7864             3606                    13
## NA19098.r1.A05     69198      8496             4381                    13
## NA19098.r1.A06     66349      8537             3235                    13
##                subsets_Mito_percent altexps_ERCC_sum altexps_ERCC_detected
##                           &lt;numeric&gt;        &lt;numeric&gt;             &lt;numeric&gt;
## NA19098.r1.A01              7.91256             1187                    31
## NA19098.r1.A02              5.99027             1277                    31
## NA19098.r1.A03              7.31765             1121                    28
## NA19098.r1.A04              6.89102             1240                    30
## NA19098.r1.A05              6.33111             1262                    33
## NA19098.r1.A06              4.87573             1308                    30
##                altexps_ERCC_percent     total
##                           &lt;numeric&gt; &lt;numeric&gt;
## NA19098.r1.A01              1.88715     62899
## NA19098.r1.A02              2.00856     63578
## NA19098.r1.A03              2.58688     43334
## NA19098.r1.A04              2.31477     53569
## NA19098.r1.A05              1.79109     70460
## NA19098.r1.A06              1.93328     67657</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(umi_feature)</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##                      mean  detected
##                 &lt;numeric&gt; &lt;numeric&gt;
## ENSG00000187634 0.0300926   2.77778
## ENSG00000188976 2.6388889  84.25926
## ENSG00000187961 0.2384259  20.60185
## ENSG00000187583 0.0115741   1.15741
## ENSG00000187642 0.0127315   1.27315
## ENSG00000188290 0.0243056   2.31481</code></pre>
<p>We can now use the functions that add the metrics calculated above to per-cell and per-gene metadata:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(umi, <span class="at">subsets=</span><span class="fu">list</span>(<span class="at">Mito=</span>is_mito))</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">&lt;-</span> <span class="fu">addPerFeatureQC</span>(umi)</span></code></pre></div>
<p>Manual filtering can use any cutoff we choose. In order to find a good value, it’s good to look at the distribution:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    umi<span class="sc">$</span>total,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">breaks =</span> <span class="dv">100</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="dv">25000</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="06a-QC_and_confounders_files/figure-html/exprs-qc7-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  umi_cell<span class="sc">$</span>detected,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">breaks =</span> <span class="dv">100</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="dv">7000</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="06a-QC_and_confounders_files/figure-html/exprs-qc8-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Sometimes it’s hard to come up with an obvious filtering cutoff. In this case, adaptive threshold can help us identify points that are more than 3 <a href="https://en.wikipedia.org/wiki/Median_absolute_deviation">median absolute deviations</a> (MADs) away from the median in any of the variables we use for QC. Be careful to specify if the correct direction of the deviation: indeed, low number of detected genes, but high MT gene percentage, are hallmarks of a low quality cell:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>qc.lib2 <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(umi_cell<span class="sc">$</span>sum, <span class="at">log=</span><span class="cn">TRUE</span>, <span class="at">type=</span><span class="st">&quot;lower&quot;</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(qc.lib2, <span class="st">&quot;thresholds&quot;</span>)</span></code></pre></div>
<pre><code>##    lower   higher 
## 23591.62      Inf</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>qc.nexprs2 <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(umi_cell<span class="sc">$</span>detected, <span class="at">log=</span><span class="cn">TRUE</span>, <span class="at">type=</span><span class="st">&quot;lower&quot;</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(qc.nexprs2, <span class="st">&quot;thresholds&quot;</span>)</span></code></pre></div>
<pre><code>##   lower  higher 
## 6259.15     Inf</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>qc.spike2 <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(umi_cell<span class="sc">$</span>altexps_ERCC_percent, <span class="at">type=</span><span class="st">&quot;higher&quot;</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(qc.spike2, <span class="st">&quot;thresholds&quot;</span>)</span></code></pre></div>
<pre><code>##    lower   higher 
##     -Inf 3.618998</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>qc.mito2 <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(umi_cell<span class="sc">$</span>subsets_Mito_percent, <span class="at">type=</span><span class="st">&quot;higher&quot;</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">attr</span>(qc.mito2, <span class="st">&quot;thresholds&quot;</span>)</span></code></pre></div>
<pre><code>##    lower   higher 
##     -Inf 9.294045</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>discard2 <span class="ot">&lt;-</span> qc.lib2 <span class="sc">|</span> qc.nexprs2 <span class="sc">|</span> qc.spike2 <span class="sc">|</span> qc.mito2</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DataFrame</span>(<span class="at">LibSize=</span><span class="fu">sum</span>(qc.lib2), <span class="at">NExprs=</span><span class="fu">sum</span>(qc.nexprs2), <span class="at">SpikeProp=</span><span class="fu">sum</span>(qc.spike2), <span class="at">MitoProp=</span><span class="fu">sum</span>(qc.mito2), <span class="at">Total=</span><span class="fu">sum</span>(discard2))</span></code></pre></div>
<pre><code>## DataFrame with 1 row and 5 columns
##     LibSize    NExprs SpikeProp  MitoProp     Total
##   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## 1        47        65       137        75       194</code></pre>
<p>All the actions performed above could be done in one <code>scater</code> command, <code>quickPerCellQC</code>:</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>reasons <span class="ot">&lt;-</span> <span class="fu">quickPerCellQC</span>(umi_cell, <span class="at">sub.fields=</span><span class="fu">c</span>(<span class="st">&quot;subsets_Mito_percent&quot;</span>, <span class="st">&quot;altexps_ERCC_percent&quot;</span>))</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">as.matrix</span>(reasons))</span></code></pre></div>
<pre><code>##              low_lib_size            low_n_features high_subsets_Mito_percent 
##                        47                        65                        75 
## high_altexps_ERCC_percent                   discard 
##                       137                       194</code></pre>
<p>Let’s add another metadata column that would keep the information about whether a cell is discarded or not:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>umi<span class="sc">$</span>discard <span class="ot">&lt;-</span> reasons<span class="sc">$</span>discard</span></code></pre></div>
<p>Plotting various coldata (cell-level medadata) assays against each other allows us to illustrate the dependencies between them. For example, cells with high mitochondrial content usually are considered dead or dying; these cells also usually have low overall UMI counts and number of detected genes.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(umi, <span class="at">x=</span><span class="st">&quot;sum&quot;</span>, <span class="at">y=</span><span class="st">&quot;subsets_Mito_percent&quot;</span>, <span class="at">colour_by=</span><span class="st">&quot;discard&quot;</span>)</span></code></pre></div>
<p><img src="06a-QC_and_confounders_files/figure-html/exprs-qc12-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(umi, <span class="at">x=</span><span class="st">&quot;sum&quot;</span>, <span class="at">y=</span><span class="st">&quot;detected&quot;</span>, <span class="at">colour_by=</span><span class="st">&quot;discard&quot;</span>)</span></code></pre></div>
<p><img src="06a-QC_and_confounders_files/figure-html/exprs-qc12-2.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(umi, <span class="at">x=</span><span class="st">&quot;altexps_ERCC_percent&quot;</span>, <span class="at">y=</span><span class="st">&quot;subsets_Mito_percent&quot;</span>,<span class="at">colour_by=</span><span class="st">&quot;discard&quot;</span>)</span></code></pre></div>
<p><img src="06a-QC_and_confounders_files/figure-html/exprs-qc12-3.png" width="100%" style="display: block; margin: auto;" /></p>
<p>We can also plot coldata with splitting by batches to see if there are substantial batch-specific differences:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(umi, <span class="at">x=</span><span class="st">&quot;sum&quot;</span>, <span class="at">y=</span><span class="st">&quot;detected&quot;</span>, <span class="at">colour_by=</span><span class="st">&quot;discard&quot;</span>, <span class="at">other_fields =</span> <span class="st">&quot;individual&quot;</span>) <span class="sc">+</span> </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>individual) <span class="sc">+</span> <span class="fu">scale_x_continuous</span>(<span class="at">labels =</span> <span class="fu">unit_format</span>(<span class="at">unit =</span> <span class="st">&quot;k&quot;</span>, <span class="at">scale =</span> <span class="fl">1e-3</span>))</span></code></pre></div>
<p><img src="06a-QC_and_confounders_files/figure-html/exprs-qc13-1.png" width="100%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(umi, <span class="at">x=</span><span class="st">&quot;sum&quot;</span>, <span class="at">y=</span><span class="st">&quot;detected&quot;</span>, <span class="at">colour_by=</span><span class="st">&quot;discard&quot;</span>, <span class="at">other_fields =</span> <span class="st">&quot;replicate&quot;</span>) <span class="sc">+</span> </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>replicate)  <span class="sc">+</span> <span class="fu">scale_x_continuous</span>(<span class="at">labels =</span> <span class="fu">unit_format</span>(<span class="at">unit =</span> <span class="st">&quot;k&quot;</span>, <span class="at">scale =</span> <span class="fl">1e-3</span>))</span></code></pre></div>
<p><img src="06a-QC_and_confounders_files/figure-html/exprs-qc13-2.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="highly-expressed-genes" class="section level3" number="1.1.4">
<h3><span class="header-section-number">1.1.4</span> Highly Expressed Genes</h3>
<p>Let’s take a look at the most expressed genes in the whole dataset. We will use symbols we obtained above. Most of the genes we see are mitochondrial or ribosomal proteins, which is pretty typical for most scRNA-seq datasets.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotHighestExprs</span>(umi, <span class="at">exprs_values =</span> <span class="st">&quot;counts&quot;</span>, </span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>                 <span class="at">feature_names_to_plot =</span> <span class="st">&quot;SYMBOL&quot;</span>, <span class="at">colour_cells_by=</span><span class="st">&quot;detected&quot;</span>)</span></code></pre></div>
<p><img src="06a-QC_and_confounders_files/figure-html/exprs-qc14-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Let’s keep the genes which were detected (expression value &gt; 1) in 2 or more cells. We’ll discard approximately 4,000 weakly expressed genes.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>keep_feature <span class="ot">&lt;-</span> <span class="fu">nexprs</span>(umi,<span class="at">byrow =</span> <span class="cn">TRUE</span>,<span class="at">detection_limit =</span> <span class="dv">1</span>) <span class="sc">&gt;=</span> <span class="dv">2</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(umi)<span class="sc">$</span>discard <span class="ot">&lt;-</span> <span class="sc">!</span> keep_feature</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">rowData</span>(umi)<span class="sc">$</span>discard)</span></code></pre></div>
<pre><code>## 
## FALSE  TRUE 
## 13882  4210</code></pre>
<p>Let’s make a new assay, <code>logcounts_raw</code>, which will contain log2-transformed counts with added pseudocount of 1.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(umi, <span class="st">&quot;logcounts_raw&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="fu">counts</span>(umi) <span class="sc">+</span> <span class="dv">1</span>)</span></code></pre></div>
<p>Finally, let’s save the <code>SingleCellExperiment</code> object with all the fields we have added to the per-cell metadata, and new assays (<code>logcounts_raw</code>):</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(umi, <span class="at">file =</span> <span class="st">&quot;data/tung/umi.rds&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="data-visualization-and-dimensionality-reduction" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Data Visualization and Dimensionality Reduction</h2>
<div id="introduction-1" class="section level3" number="1.2.1">
<h3><span class="header-section-number">1.2.1</span> Introduction</h3>
<p>In this chapter we will continue to work with the filtered <code>Tung</code> dataset produced in the previous chapter. We will explore different ways of visualizing the data to allow you to asses what happened to the expression matrix after the quality control step. <code>scater</code> package provides several very useful functions to simplify visualisation.</p>
<p>One important aspect of single-cell RNA-seq is to control for batch effects. Batch effects are technical artefacts that are added to the samples during handling. For example, if two sets of samples were prepared in different labs or even on different days in the same lab, then we may observe greater similarities between the samples that were handled together. In the worst case scenario, batch effects may be <a href="http://f1000research.com/articles/4-121/v1">mistaken</a> for true biological variation. The <code>Tung</code> data allows us to explore these issues in a controlled manner since some of the salient aspects of how the samples were handled have been recorded. Ideally, we expect to see batches from the same individual grouping together and distinct groups corresponding to each individual.</p>
<p>Let’s create another <code>SingleCellExperiment</code> object, <code>umi.qc</code>, in which remove unnecessary poorly expressed genes and low quality cells.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>umi.qc <span class="ot">&lt;-</span> umi[<span class="sc">!</span> <span class="fu">rowData</span>(umi)<span class="sc">$</span>discard,<span class="sc">!</span> <span class="fu">colData</span>(umi)<span class="sc">$</span>discard]</span></code></pre></div>
</div>
<div id="pca-plot" class="section level3" number="1.2.2">
<h3><span class="header-section-number">1.2.2</span> PCA plot</h3>
<p>The easiest way to overview the data is by transforming it using the principal component analysis and then visualize the first two principal components.</p>
<p><a href="https://en.wikipedia.org/wiki/Principal_component_analysis">Principal component analysis (PCA)</a> is a statistical procedure that uses a transformation to convert a set of observations into a set of linearly uncorrelated (orthogonal) variables called principal components (PCs). The number of principal components is less than or equal to the number of original variables.</p>
<p>Mathematically, the PCs correspond to the <a href="https://en.wikipedia.org/wiki/Eigenvalues_and_eigenvectors">eigenvectors</a> of the covariance matrix. The eigenvectors are sorted by eigenvalue so that the first principal component accounts for as much of the variability in the data as possible, and each succeeding component in turn has the highest variance possible under the constraint that it is orthogonal to the preceding components (the figure below is taken from <a href="http://www.nlpca.org/pca_principal_component_analysis.html">here</a>).</p>
<div class="figure" style="text-align: center"><span id="fig:exprs-over2"></span>
<img src="figures/pca.png" alt="Schematic representation of PCA dimensionality reduction" width="100%" />
<p class="caption">
Figure 1.1: Schematic representation of PCA dimensionality reduction
</p>
</div>
<div id="before-qc" class="section level4" number="1.2.2.1">
<h4><span class="header-section-number">1.2.2.1</span> Before QC</h4>
<p>Without log-transformation or normalization, PCA plot fails to separate the datasets by replicate or individual. We mostly see the effects of sequencing depth - samples (cells) with lots of expression, and particularly highly expressed genes, dominate the PCs:</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(umi, <span class="at">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">reducedDim</span>(umi, <span class="st">&quot;PCA&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 864  50</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(umi, <span class="at">colour_by =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">size_by =</span> <span class="st">&quot;detected&quot;</span>, <span class="at">shape_by =</span> <span class="st">&quot;individual&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:exprs-over3"></span>
<img src="06a-QC_and_confounders_files/figure-html/exprs-over3-1.png" alt="PCA plot of the Tung data (raw counts)" width="100%" />
<p class="caption">
Figure 1.2: PCA plot of the Tung data (raw counts)
</p>
</div>
<p>With log-transformation, we equalize the large difference between strongly and weakly expressed genes, and immediately see cells form groups by replicate, individual, and sequencing depth. When PCA is re-run, reducedDim object in <code>umi</code> is overwritten.</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(umi, <span class="at">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">reducedDim</span>(umi, <span class="st">&quot;PCA&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 864  50</code></pre>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(umi, <span class="at">colour_by =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">size_by =</span> <span class="st">&quot;detected&quot;</span>, <span class="at">shape_by =</span> <span class="st">&quot;individual&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:exprs-over4"></span>
<img src="06a-QC_and_confounders_files/figure-html/exprs-over4-1.png" alt="PCA plot of the tung data (non-normalized logcounts)" width="100%" />
<p class="caption">
Figure 1.3: PCA plot of the tung data (non-normalized logcounts)
</p>
</div>
<p>Clearly log-transformation is benefitial for our data - it reduces the variance on the first principal component and already separates some biological effects. Moreover, it makes the distribution of the expression values more normal. In the following analysis and chapters we will be using log-transformed raw counts by default.</p>
<p><strong>However, note that just a log-transformation is not enough to account for different technical factors between the cells (e.g. sequencing depth). Therefore, please do not use <code>logcounts_raw</code> for your downstream analysis, instead as a minimum suitable data use the <code>logcounts</code> slot of the <code>SingleCellExperiment</code> object, which not just log-transformed, but also normalised by library size (e.g. CPM normalisation). In the course we use <code>logcounts_raw</code> only for demonstration purposes!</strong></p>
</div>
<div id="after-qc" class="section level4" number="1.2.2.2">
<h4><span class="header-section-number">1.2.2.2</span> After QC</h4>
<p>Let’s do the same analysis as above, but using <code>umi.qc</code> dataframe instead of the full <code>umi</code>:</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>umi.qc <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(umi.qc, <span class="at">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">reducedDim</span>(umi.qc, <span class="st">&quot;PCA&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 670  50</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(umi.qc, <span class="at">colour_by =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">size_by =</span> <span class="st">&quot;detected&quot;</span>, <span class="at">shape_by =</span> <span class="st">&quot;individual&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:exprs-over5"></span>
<img src="06a-QC_and_confounders_files/figure-html/exprs-over5-1.png" alt="PCA plot of the Tung data (non-normalized log counts, QC-filtered)" width="100%" />
<p class="caption">
Figure 1.4: PCA plot of the Tung data (non-normalized log counts, QC-filtered)
</p>
</div>
<p>Comparing figures above, it is clear that after quality control the NA19098.r2 cells no longer form a group of outliers.</p>
<p>By default only the top 500 most variable genes are used by <code>scater</code> to calculate the PCA. This can be adjusted by changing the <code>ntop</code> argument.</p>
<p><strong>Exercise 1</strong>
How do the PCA plots change if when all 14,154 genes are used? Or when only top 50 genes are used? Why does the fraction of variance accounted for by the first PC change so dramatically?</p>
<p><strong>Hint</strong> Use <code>ntop</code> argument of the <code>plotPCA</code> function.</p>
<details>
<summary>
Answer
</summary>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>umi.qc <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(umi.qc, <span class="at">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,<span class="at">ntop =</span> <span class="fu">nrow</span>(umi.qc))</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(umi.qc, <span class="at">colour_by =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">size_by =</span> <span class="st">&quot;detected&quot;</span>, <span class="at">shape_by =</span> <span class="st">&quot;individual&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:exprs-over6"></span>
<img src="06a-QC_and_confounders_files/figure-html/exprs-over6-1.png" alt="PCA plot of the tung data (14214 genes)" width="100%" />
<p class="caption">
Figure 1.5: PCA plot of the tung data (14214 genes)
</p>
</div>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>umi.qc <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(umi.qc, <span class="at">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,<span class="at">ntop =</span> <span class="dv">50</span>)</span></code></pre></div>
<pre><code>## Warning in check_numbers(k = k, nu = nu, nv = nv, limit = min(dim(x)) - : more
## singular values/vectors requested than available</code></pre>
<pre><code>## Warning in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth =
## TRUE, : You&#39;re computing too large a percentage of total singular values, use a
## standard svd instead.</code></pre>
<pre><code>## Warning in (function (A, nv = 5, nu = nv, maxit = 1000, work = nv + 7, reorth
## = TRUE, : did not converge--results might be invalid!; try increasing work or
## maxit</code></pre>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(umi.qc, <span class="at">colour_by =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">size_by =</span> <span class="st">&quot;detected&quot;</span>, <span class="at">shape_by =</span> <span class="st">&quot;individual&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="06a-QC_and_confounders_files/figure-html/exprs-over7,%20expr-overview-pca-after-qc-exercise1-2-1.png" alt="PCA plot of the tung data (50 genes)" width="100%" />
<p class="caption">
(#fig:exprs-over7, expr-overview-pca-after-qc-exercise1-2)PCA plot of the tung data (50 genes)
</p>
</div>
If your answers are different please compare your code with <a href="https://github.com/hemberg-lab/scRNA.seq.course/blob/master/07-exprs-overview.Rmd">ours</a> (you need to search for this exercise in the opened file).
</details>
</div>
</div>
<div id="tsne-map" class="section level3" number="1.2.3">
<h3><span class="header-section-number">1.2.3</span> tSNE Map</h3>
<p>An alternative to PCA for visualizing scRNA-seq data is a tSNE plot. <a href="https://lvdmaaten.github.io/tsne/">tSNE</a> (t-Distributed Stochastic Neighbor Embedding) combines dimensionality reduction (e.g. PCA) with random walks on the nearest-neighbour network to map high dimensional data (i.e. our 14,154-dimensional expression matrix) to a 2-dimensional space while preserving local distances between cells. In contrast with PCA, tSNE is a stochastic algorithm which means running the method multiple times on the same dataset will result in different plots. Due to the non-linear and stochastic nature of the algorithm, tSNE is more difficult to intuitively interpret tSNE. To ensure reproducibility, we fix the “seed” of the random-number generator in the code below so that we always get the same plot.</p>
<div id="before-qc-1" class="section level4" number="1.2.3.1">
<h4><span class="header-section-number">1.2.3.1</span> Before QC</h4>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123456</span>)</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>umi <span class="ot">&lt;-</span> <span class="fu">runTSNE</span>(umi, <span class="at">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>, <span class="at">perplexity =</span> <span class="dv">130</span>)</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTSNE</span>(umi, <span class="at">colour_by =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">size_by =</span> <span class="st">&quot;detected&quot;</span>, <span class="at">shape_by =</span> <span class="st">&quot;individual&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:exprs-over8"></span>
<img src="06a-QC_and_confounders_files/figure-html/exprs-over8-1.png" alt="tSNE map of the tung data" width="100%" />
<p class="caption">
Figure 1.6: tSNE map of the tung data
</p>
</div>
</div>
<div id="after-qc-1" class="section level4" number="1.2.3.2">
<h4><span class="header-section-number">1.2.3.2</span> After QC</h4>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123456</span>)</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>umi.qc <span class="ot">&lt;-</span> <span class="fu">runTSNE</span>(umi.qc, <span class="at">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>, <span class="at">perplexity =</span> <span class="dv">130</span>)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotTSNE</span>(umi.qc, <span class="at">colour_by =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">size_by =</span> <span class="st">&quot;detected&quot;</span>, <span class="at">shape_by =</span> <span class="st">&quot;individual&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:exprs-over9"></span>
<img src="06a-QC_and_confounders_files/figure-html/exprs-over9-1.png" alt="tSNE map of the tung data" width="100%" />
<p class="caption">
Figure 1.7: tSNE map of the tung data
</p>
</div>
<p>Interpreting PCA and tSNE plots is often challenging and due to their stochastic and non-linear nature, they are less intuitive. However, in this case it is clear that they provide a similar picture of the data. Comparing figures above, it is again clear that the samples from NA19098.r2 are no longer outliers after the QC filtering.</p>
<p>Furthermore tSNE requires you to provide a value of <code>perplexity</code> which reflects the number of neighbours used to build the nearest-neighbour network; a high value creates a dense network which clumps cells together while a low value makes the network more sparse allowing groups of cells to separate from each other. <code>scater</code> uses a default perplexity of the total number of cells divided by five (rounded down).</p>
<p>You can read more about the pitfalls of using tSNE <a href="http://distill.pub/2016/misread-tsne/">here</a>. A more recent publication entitled <a href="https://www.nature.com/articles/s41467-019-13056-x">“The art of using t-SNE for single-cell transcriptomics”</a> discusses similarities and differences between t-SNE and UMAP, finding that most observed differences are due to initialization, and gives recommendataion on parameter tuning when visualizing scRNA-seq datasets of different sizes.</p>
<p><strong>Exercise 2</strong>
How do the tSNE plots change when a perplexity of 10 or 200 is used? How does the choice of perplexity affect the interpretation of the results?</p>
<details>
<summary>
Answer
</summary>
<div class="figure" style="text-align: center"><span id="fig:exprs-over10"></span>
<img src="06a-QC_and_confounders_files/figure-html/exprs-over10-1.png" alt="tSNE map of the tung data (perplexity = 10)" width="100%" />
<p class="caption">
Figure 1.8: tSNE map of the tung data (perplexity = 10)
</p>
</div>
<div class="figure" style="text-align: center"><span id="fig:exprs-over11"></span>
<img src="06a-QC_and_confounders_files/figure-html/exprs-over11-1.png" alt="tSNE map of the tung data (perplexity = 200)" width="100%" />
<p class="caption">
Figure 1.9: tSNE map of the tung data (perplexity = 200)
</p>
</div>
</details>
</div>
</div>
</div>
<div id="identifying-confounding-factors" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> Identifying Confounding Factors</h2>
<div id="introduction-2" class="section level3" number="1.3.1">
<h3><span class="header-section-number">1.3.1</span> Introduction</h3>
<p>There is a large number of potential confounders, artifacts and biases in scRNA-seq data. One of the main challenges in analyzing scRNA-seq data stems from the fact that it is difficult to carry out a true technical replicate (why?) to distinguish biological and technical variability. In the previous chapters we considered batch effects and in this chapter we will continue to explore how experimental artifacts can be identified and removed. We will continue using the <code>scater</code> package since it provides a set of methods specifically for quality control of experimental and explanatory variables. Moreover, we will continue to work with the Blischak data that was used in the previous chapter.</p>
<p>Our <code>umi.qc</code> dataset contains filtered cells and genes. Our next step is to explore technical drivers of variability in the data to inform data normalisation before downstream analysis.</p>
</div>
<div id="correlations-with-pcs" class="section level3" number="1.3.2">
<h3><span class="header-section-number">1.3.2</span> Correlations with PCs</h3>
<p>Let’s first look again at the PCA plot of the QC-filtered dataset:</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>umi.qc <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(umi.qc, <span class="at">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>)</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">reducedDim</span>(umi.qc, <span class="st">&quot;PCA&quot;</span>))</span></code></pre></div>
<pre><code>## [1] 670  50</code></pre>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(umi.qc, <span class="at">colour_by =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">size_by =</span> <span class="st">&quot;sum&quot;</span>, <span class="at">shape_by =</span> <span class="st">&quot;individual&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:confounders2"></span>
<img src="06a-QC_and_confounders_files/figure-html/confounders2-1.png" alt="PCA plot of the tung data" width="100%" />
<p class="caption">
Figure 1.10: PCA plot of the tung data
</p>
</div>
<p><code>scater</code> allows one to identify principal components that correlate with experimental and QC variables of interest (it ranks principle components by <span class="math inline">\(R^2\)</span> from a linear model regressing PC value against the variable of interest).</p>
<p>Let’s test whether some of the variables correlate with any of the PCs.</p>
<div id="detected-genes" class="section level4" number="1.3.2.1">
<h4><span class="header-section-number">1.3.2.1</span> Detected genes</h4>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">logcounts</span>(umi.qc) <span class="ot">&lt;-</span> <span class="fu">assay</span>(umi.qc, <span class="st">&quot;logcounts_raw&quot;</span>)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="fu">getExplanatoryPCs</span>(umi.qc,<span class="at">variables =</span> <span class="st">&quot;sum&quot;</span>)</span></code></pre></div>
<pre><code>##               sum
## PC1  8.664160e+01
## PC2  5.688940e+00
## PC3  1.709397e-01
## PC4  4.581186e-01
## PC5  1.813745e-04
## PC6  6.802961e-01
## PC7  1.058893e-01
## PC8  2.043980e-01
## PC9  5.650720e-01
## PC10 4.694972e-03</code></pre>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotExplanatoryPCs</span>(umi.qc,<span class="at">variables =</span> <span class="st">&quot;sum&quot;</span>) </span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:confounders3"></span>
<img src="06a-QC_and_confounders_files/figure-html/confounders3-1.png" alt="PC correlation with the number of detected genes" width="100%" />
<p class="caption">
Figure 1.11: PC correlation with the number of detected genes
</p>
</div>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">logcounts</span>(umi.qc) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span></code></pre></div>
<p>Indeed, we can see that <code>PC1</code> can be almost completely (86%) explained by the total UMI counts (sequencing depth). In fact, it was also visible on the PCA plot above. This is a well-known issue in scRNA-seq and was described <a href="http://biorxiv.org/content/early/2015/12/27/025528">here</a>.</p>
</div>
</div>
<div id="explanatory-variables" class="section level3" number="1.3.3">
<h3><span class="header-section-number">1.3.3</span> Explanatory Variables</h3>
<p><code>scater</code> can also compute the marginal <span class="math inline">\(R^2\)</span> for each variable when fitting a linear model regressing expression values for each gene against just that variable, and display a density plot of the gene-wise marginal <span class="math inline">\(R^2\)</span> values for the variables.</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotExplanatoryVariables</span>(umi.qc,<span class="at">exprs_values =</span> <span class="st">&quot;logcounts_raw&quot;</span>,</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">variables =</span> <span class="fu">c</span>(<span class="st">&quot;detected&quot;</span>,<span class="st">&quot;sum&quot;</span>,<span class="st">&quot;batch&quot;</span>,</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>                                       <span class="st">&quot;individual&quot;</span>,<span class="st">&quot;altexps_ERCC_percent&quot;</span>,<span class="st">&quot;subsets_Mito_percent&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:confounders4"></span>
<img src="06a-QC_and_confounders_files/figure-html/confounders4-1.png" alt="Explanatory variables" width="100%" />
<p class="caption">
Figure 1.12: Explanatory variables
</p>
</div>
<p>This analysis indicates that the number of detected genes (again) and also the sequencing depth (number of counts) have substantial explanatory power for many genes, so these variables are good candidates for conditioning out in a normalisation step, or including in downstream statistical models. Expression of ERCCs also appears to be an important explanatory variable and one notable feature of the above plot is that batch explains more than individual. What does that tell us about the technical and biological variability of the data?</p>
</div>
<div id="other-confounders" class="section level3" number="1.3.4">
<h3><span class="header-section-number">1.3.4</span> Other Confounders</h3>
<p>In addition to correcting for batch, there are other factors that one may want to compensate for. As with batch correction, these adjustments require extrinsic information. One popular method is <a href="https://github.com/PMBio/scLVM">scLVM</a> which allows you to identify and subtract the effect from processes such as cell-cycle or apoptosis.</p>
<p>In addition, protocols may differ in terms of their coverage of each transcript, their bias based on the average content of <strong>A/T</strong> nucleotides, or their ability to capture short transcripts. Ideally, we would like to compensate for all of these differences and biases.</p>
</div>
<div id="exercise" class="section level3" number="1.3.5">
<h3><span class="header-section-number">1.3.5</span> Exercise</h3>
<p>Perform the same analysis with read counts of the Blischak data. Use <code>tung/reads.rds</code> file to load the reads SCESet object. Once you have finished please compare your results to ours (next chapter).</p>
</div>
<div id="sessioninfo" class="section level3" number="1.3.6">
<h3><span class="header-section-number">1.3.6</span> sessionInfo()</h3>
<details>
<summary>
View session info
</summary>
<pre><code>## R version 4.0.4 (2021-02-15)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] scales_1.1.1                EnsDb.Hsapiens.v86_2.99.0  
##  [3] ensembldb_2.14.1            AnnotationFilter_1.14.0    
##  [5] GenomicFeatures_1.42.2      org.Hs.eg.db_3.12.0        
##  [7] AnnotationDbi_1.52.0        scater_1.18.6              
##  [9] ggplot2_3.3.4               SingleCellExperiment_1.12.0
## [11] SummarizedExperiment_1.20.0 Biobase_2.50.0             
## [13] GenomicRanges_1.42.0        GenomeInfoDb_1.26.4        
## [15] IRanges_2.24.1              S4Vectors_0.28.1           
## [17] BiocGenerics_0.36.0         MatrixGenerics_1.2.1       
## [19] matrixStats_0.58.0          knitr_1.31                 
## 
## loaded via a namespace (and not attached):
##  [1] Rtsne_0.15                ggbeeswarm_0.6.0         
##  [3] colorspace_2.0-0          ellipsis_0.3.1           
##  [5] scuttle_1.0.4             XVector_0.30.0           
##  [7] BiocNeighbors_1.8.2       farver_2.1.0             
##  [9] bit64_4.0.5               fansi_0.4.2              
## [11] xml2_1.3.2                codetools_0.2-18         
## [13] sparseMatrixStats_1.2.1   cachem_1.0.4             
## [15] jsonlite_1.7.2            Rsamtools_2.6.0          
## [17] dbplyr_2.1.0              compiler_4.0.4           
## [19] httr_1.4.2                assertthat_0.2.1         
## [21] Matrix_1.3-2              fastmap_1.1.0            
## [23] lazyeval_0.2.2            BiocSingular_1.6.0       
## [25] htmltools_0.5.1.1         prettyunits_1.1.1        
## [27] tools_4.0.4               rsvd_1.0.3               
## [29] gtable_0.3.0              glue_1.4.2               
## [31] GenomeInfoDbData_1.2.4    dplyr_1.0.5              
## [33] rappdirs_0.3.3            Rcpp_1.0.6               
## [35] jquerylib_0.1.3           vctrs_0.3.6              
## [37] Biostrings_2.58.0         rtracklayer_1.50.0       
## [39] DelayedMatrixStats_1.12.3 xfun_0.22                
## [41] stringr_1.4.0             beachmat_2.6.4           
## [43] lifecycle_1.0.0           irlba_2.3.3              
## [45] XML_3.99-0.6              zlibbioc_1.36.0          
## [47] hms_1.0.0                 ProtGenerics_1.22.0      
## [49] yaml_2.2.1                curl_4.3                 
## [51] memoise_2.0.0             gridExtra_2.3            
## [53] sass_0.3.1                biomaRt_2.46.3           
## [55] stringi_1.5.3             RSQLite_2.2.4            
## [57] highr_0.8                 BiocParallel_1.24.1      
## [59] rlang_0.4.10              pkgconfig_2.0.3          
## [61] bitops_1.0-6              evaluate_0.14            
## [63] lattice_0.20-41           purrr_0.3.4              
## [65] GenomicAlignments_1.26.0  labeling_0.4.2           
## [67] cowplot_1.1.1             bit_4.0.4                
## [69] tidyselect_1.1.0          magrittr_2.0.1           
## [71] bookdown_0.22             R6_2.5.0                 
## [73] generics_0.1.0            DelayedArray_0.16.3      
## [75] DBI_1.1.1                 pillar_1.5.1             
## [77] withr_2.4.1               RCurl_1.98-1.3           
## [79] tibble_3.1.0              crayon_1.4.1             
## [81] utf8_1.2.1                BiocFileCache_1.14.0     
## [83] rmarkdown_2.7             viridis_0.5.1            
## [85] progress_1.2.2            grid_4.0.4               
## [87] blob_1.2.1                digest_0.6.27            
## [89] openssl_1.4.3             munsell_0.5.0            
## [91] beeswarm_0.3.1            viridisLite_0.3.0        
## [93] vipor_0.4.5               bslib_0.2.4              
## [95] askpass_1.1</code></pre>
</details>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>


    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"search": false
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
